### PR TITLE
Gaussian probe

### PR DESCRIPTION
**Pull request by fym0503**

### CODE REVIEW

File: neurox/interpretation/gaussian_probe.py [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+            ].mean()
+        mutual_inf = (entropy - conditional_entropy) / torch.tensor(2.0).log()
+        return accuracy, mutual_inf.item(), (mutual_inf / entropy).item()
+
+
+def train_probe(X, y):
+    return GaussianProbe(X, y)
+
+
+def evaluate_probe(probe, X_test, y_test, selected_neurons):
```
Comment (fdalvi) [inline neurox/interpretation/gaussian_probe.py:146]: Is it possible to "evaluate" a probe on all neurons? If yes we can make `selected_neurons` optional with the default behavior being evaluation on all neurons

File: neurox/interpretation/gaussian_probe.py [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+
+    probe.test_categorical = probe._get_categorical("test")
+
+    probe.categorical_sets["test"] = probe.test_categorical
+
+    probe._get_distributions(selected_neurons)
+
+    probe._compute_probs(selected_neurons, "test")
+    test_acc, test_mi, test_nmi = probe._predict("test")
+    return test_acc
```
Comment (fdalvi) [inline neurox/interpretation/gaussian_probe.py:161]: Can we add a `return_predictions` argument like linear_probe (see https://neurox.qcri.org/docs/neurox.interpretation.html#neurox.interpretation.linear_probe.evaluate_probe). It has been quite useful for us in the past

File: neurox/interpretation/gaussian_probe.py [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+    probe.label_sets["test"] = probe.test_labels
+
+    probe.test_categorical = probe._get_categorical("test")
+
+    probe.categorical_sets["test"] = probe.test_categorical
+
+    probe._get_distributions(selected_neurons)
+
+    probe._compute_probs(selected_neurons, "test")
+    test_acc, test_mi, test_nmi = probe._predict("test")
```
Comment (fdalvi) [inline neurox/interpretation/gaussian_probe.py:160]: the linear_probe also supports various metrics other than accuracy (see https://neurox.qcri.org/docs/neurox.interpretation.html#neurox.interpretation.linear_probe.evaluate_probe), do you think its useful to implement that here as well?

Keep in mind that we want to provide a "general" implementation that people can use for their own tasks, so we don't need to exactly stick to the original implementation (although it is a good idea to default to the original implementation)

File: neurox/interpretation/gaussian_probe.py [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    return test_acc
+
+
+def get_neuron_ordering(probe, num_of_neurons):
+    selected_neurons = []
+    for num_of_neuron in range(len(selected_neurons), num_of_neurons):
```
Comment (fdalvi) [inline neurox/interpretation/gaussian_probe.py:166]: `len(selected_neurons)` always evaluates to 0, so just use `range(num_of_neurons)`

File: neurox/interpretation/gaussian_probe.py [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+            acc, mi, nmi = probe._predict("train")
+
+            if mi > best_mi:
+                best_mi = mi
+                best_nmi = nmi
+                best_neuron = neuron
+                acc_on_best_mi = acc
+
+        selected_neurons.append(best_neuron)
+        print("added neuron ", best_neuron)
```
Comment (fdalvi) [inline neurox/interpretation/gaussian_probe.py:187]: remove spurious `print`

File: neurox/interpretation/gaussian_probe.py [hunk 6]
```diff
@@ -0,0 +1,191 @@
...
+        probe._get_distributions(selected_neurons)
```
Comment (fdalvi) [inline neurox/interpretation/gaussian_probe.py:188]: any reason we are doing the computation on lines 188-190? I don't see it being used anywhere. If thats the case please remove the lines

File: tests/interpretation/test_gaussian_probe.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+            [
+                [-1.1, -1.05, 0.01],
+                [-1.05, 0, 0.005],
+                [1.1, 1.05, 0.003],
+                [1.03, 0, 0.1],
+            ],
+            dtype=float,
+        )
+        y = np.array([0, 0, 1, 1])
+        expected_neuron_order = [0, 1, 2]
```
Comment (fdalvi) [inline tests/interpretation/test_gaussian_probe.py:27]: Can you rearrange the data such that the neuron order is _not_ `0,1,2`; in case the implementation always picks the "next available neuron", this test will not distinguish it from a working implementation;

**Review by fdalvi — CHANGES_REQUESTED** [review]
Looks really good @fym0503, I think we only need a few minor changes. I've left some comments. Two overarching comments:

1. Please add docstrings for the public facing functions (feel free to use https://github.com/fdalvi/NeuroX/blob/d277515e659f66d7f272155fdfa22f95309ca9f6/neurox/interpretation/linear_probe.py#L218 as a starting point)
2. I think you continued working on your old branch (this PR re-adds IoU probe stuff). To fix this, you can:
```bash
git checkout master
git checkout -b feat/gaussian_probe
git cherry-pick 56f24ad4b086d6d6677124962ef1a27af0264beb 7964ac50c8dfd703f861a56b5479e5b2c0907fdd ad02325db2da693be44d58b18145f7889e743ef1
git push origin feat/gaussian_probe
```

Now you will have a "fixed" branch `feat/gaussian_probe` without any of the older IoU probe changes. You can change the `base_branch` of this PR through GitHub's UI to `feat/gaussian_probe` and continue pushing to that branch.

Please don't hesitate to reach out if you need any help. Good work!

**Comment by fdalvi** [comment]
Closing this PR as its continued in #30
